libbpf-rs/src/object.rs

```diff
@@ -266,12 +266,10 @@ impl OpenObject {
             return Err(Error::System(-ret));
         }
 
-        let obj = Object::new(self.ptr)?;
-
+        let ptr = self.ptr;
         // Prevent object from being closed once `self` is dropped
         self.ptr = ptr::null_mut();
-
-        Ok(obj)
+        Object::new(ptr)
```

Comment:
I don't see double-free on the original code and your changes would introduce a leak.

Reply:
OK, I see the problem. We can prevent create `Object` prematurely in Object::new() instead.

Reply:
So you want `Object::new` to take ownership of the input pointer in one codepath, but not the other one. It's kinda ambiguous, don't you think? That's exactly the reason there is a bug here. Could you care to explain why is it better than the solution I provided?

Reply:
I think your current approach here @trtt is the most straightforward one / simplest fix to me.